src/bin/ipa_bench/gen_events.rs

```diff
@@ -0,0 +1,481 @@
+}
+
+// TODO: Currently, users are mutually exclusive in each ad loop (i.e. User A in ad X will never appear in other ads).
+// We need to generate events from same users across ads (but how often should a user appear in different ads?)
+// "Ads" doesn't mean FB's L3 ads. It could be ads from different businesses.
+
+pub fn generate_events<W: io::Write>(
+    sample: &Sample,
+    total_count: u32,
+    epoch: u8,
+    secret_share: bool,
+    seed: &Option<u64>,
+    out: &mut W,
+) -> (u32, u32) {
+    let mut rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
+    };
+    debug!("seed: {:?}", rng.get_seed());
+
+    // Separate RNG for generating secret shares
+    let mut ss_rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
+    };
+
+    let mut ad_count = 0;
+    let mut event_count = 0;
+    let mut total_impressions = 0;
+    let mut total_conversions = 0;
+
+    // Simulate impressions and conversions from an ad.
+    // We define "ad" as a group of impressions and conversions from targeted users who are selected by predefined
+    // breakdowns such as age, gender and locations.
+    loop {
+        ad_count += 1;
+        debug!("ad: {}", ad_count);
+
+        // For now, we'll do 1 ad = 1 breakdown key
+        let ad_id: u32 = rng.gen();
+
+        // Number of unique people who saw the ad
+        let reach = sample.reach_per_ad(&mut rng);
+        debug!("reach: {}", reach);
+
+        // CVR for the ad
+        let cvr = sample.cvr_per_ad_account(&mut rng);
+        debug!("CVR: {}", cvr);
+
+        for _ in 0..reach {
+            // # of devices == # of matchkeys
+            let devices = sample.devices_per_user(&mut rng);
+            trace!("devices per user: {}", devices);
+
+            let impressions = sample.impression_per_user(&mut rng);
+            trace!("impressions per user: {}", impressions);
+
+            // Probabilistically decide whether this user has converted or not
+            let conversions = if Bernoulli::new(cvr).unwrap().sample(&mut rng) {
+                sample.conversion_per_user(&mut rng)
+            } else {
+                0
+            };
+            trace!("conversions per user: {}", conversions);
+
+            let events = gen_events(
+                &GenEventParams {
+                    devices,
+                    impressions,
+                    conversions,
+                    epoch,
+                    breakdown_key: ad_id.to_string(),
+                },
+                secret_share,
+                sample,
+                &mut rng,
+                &mut ss_rng,
+            );
+
+            total_impressions += impressions.to_u32().unwrap();
+            total_conversions += conversions.to_u32().unwrap();
+
+            for e in events {
+                out.write_all(serde_json::to_string(&e).unwrap().as_bytes())
```

Comment:
How are events delineated in the file?  It looks like you have an EOL on each, but that relies on serde_json not adding an EOL.  Try ASCII RS instead: https://datatracker.ietf.org/doc/html/rfc7464